pallets/accounts/src/lib.rs

```diff
@@ -507,57 +603,121 @@ pub mod pallet {
-						}
-					}
-				}
+					// mint Tickets for the linked user
+					<pallet_assets::Pallet<T> as Mutate<T::AccountId>>::mint_into(
+						T::TicketsAssetId::get(),
+						&linked,
+						tickets_amount, // amount already ensured to be > 0 in case of Lock
+					)?;
+					// assign NOVA
+					let nova_old_balance = pallet_balances::Pallet::<T>::free_balance(&linked);
+					ensure!(
+						nova_old_balance + nova_amount >=
+							<pallet_balances::Pallet<T> as Currency<T::AccountId>>::minimum_balance(
+							),
+						Error::<T>::BelowMinimumBalance
+					);
+					ensure!(
+						!nova_old_balance.checked_add(&nova_amount).is_none(),
+						Error::<T>::SystematicFailure
+					);
 
-				// also emit event
-				Self::deposit_event(Event::Locked { eth_key: sender, balance: amount, locktime });
-			// 问Gio for clarification
-			} else {
-				// If we want to unlock all the FRAG tokens that were
-				// ! TODO TEST
+					// Checks passed. Now write.
+					let _ =
+						<pallet_balances::Pallet<T> as fungible::Mutate<T::AccountId>>::mint_into(
+							&linked,
+							nova_amount, // amount already ensured to be > 0 in case of Lock
+						)?;
 
-				// if we have any link to this account, then force unlinking
-				let linked = <EVMLinksReverse<T>>::get(sender.clone());
-				if let Some(linked) = linked {
-					Self::unlink_account(linked, sender.clone())?; // Unlink Ethereum Account `sender` and Clamor Account `linked`
+					Self::deposit_event(TicketsMinted {
+						sender: linked.clone(),
+						balance: tickets_amount,
+					});
+					Self::deposit_event(NOVAAssigned {
+						sender: linked.clone(),
+						balance: nova_amount,
+					});
+				} else {
+					// Ethereum Account ID (H160) not linked to Clamor Account ID
+					// So, register the amount of tickets and NOVA owned by the H160 account for later linking
+					<EthReservedTickets<T>>::insert(sender.clone(), tickets_amount);
+					<EthReservedNova<T>>::insert(sender.clone(), nova_amount);
+
+					Self::deposit_event(TicketsReserved {
+						eth_key: sender.clone(),
+						balance: tickets_amount,
+					});
+					Self::deposit_event(NOVAReserved {
+						eth_key: sender.clone(),
+						balance: nova_amount,
+					});
 				}
-
 				// also emit event
-				Self::deposit_event(Event::Unlocked { eth_key: sender, balance: amount }); // 问Gio for clarification
+				Self::deposit_event(Event::Locked {
+					eth_key: sender,
+					balance: frag_amount,
+					lock_period,
+				});
+			} else {
+				Self::deposit_event(Event::Unlocked { eth_key: sender, balance: frag_amount });
 			}
 
-			// write this later as unlink_account can fail
 			<EthLockedFrag<T>>::insert(
-				// VERY IMPORTANT TO NOTE
 				sender.clone(),
-				EthLock { amount, block_number: current_block_number },
+				EthLock {
+					amount: frag_amount, // amount already ensured to be > 0 for lock, = 0 for unlock
+					block_number: current_block_number,
+					first_lock_block_number: first_lock_block_num,
```

Comment:
superfluous no? the above is the exact same value

Reply:
not the same. Explained before. Lock and Unlock updates the `block_number` field.

Reply:
I now split the Lock from the Unlock, creating a new type [EthUnlockedFrag](https://github.com/fragcolor-xyz/clamor/blob/6fe5df92349f589002381c55d56f39a6e90ec426/pallets/accounts/src/lib.rs#L289). Having both using the same storage was complicating things in the withdraw calculations.